plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ScopeIndex.kt

```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright 2010-2025 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.conversion
+
+/**
+ * All the scopes in the method are indexed with some integer.
+ * This integer is then used in names of locals in that scope as Viper does not allow shadowing
+ * and each variable must have a unique identifier.
+ * In some cases, however, we create a separate context, but it does not correspond to a scope.
+ * We use `NoScope` there.
```

Comment:
I think this reformulation may be better:
```suggestion
 * When converting Kotlin methods, we explicitly track what scope local variables are declared in
 * to resolve shadowing. In contexts where local variables are permitted, we use a special `NoScope`
 * placeholder that is treated as an error downstream.
```